File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part05_bestSum/java/BestSum.java
import java.util.ArrayList;

public class BestSum {
  public static void main(String[] args) {
    System.out.println(doBestSum(7, new int[]{5, 3, 4, 7}));      //true
    System.out.println(doBestSum(8, new int[]{2, 3, 5}));  //true
    System.out.println(doBestSum(8, new int[]{1, 4, 5}));      //false
    System.out.println(doBestSum(25, new int[]{1, 2, 5, 25}));    //true
    System.out.println(doBestSum(0, new int[]{7,14}));   //false
  }

  public static ArrayList<Integer> doBestSum(int targetSum, int[] numbers){
    if(targetSum == 0) return new ArrayList<Integer>();
    if(targetSum < 0) return null;

    ArrayList<Integer> shortestCombination = null;

    for( int num : numbers){
      int remainder = targetSum - num;
      ArrayList<Integer> remainderCombination = doBestSum(remainder, numbers);

      if(remainderCombination != null){
        remainderCombination.add(num);

        //ArrayList<Integer> combination = remainderCombination;

        if(shortestCombination == null || remainderCombination.size() < shortestCombination.size()){
          shortestCombination = remainderCombination;
        }

      }
    }
    
    return shortestCombination;
  }
}

File: data_structures_and_algorithms/alvin-zablan/dynamic-programming/part05_bestSum/javascript/bestSum.js
const bestSum = (targetSum, numbers) => {
  if (targetSum === 0) return [];
  if (targetSum < 0) return null;

  let shortestCombination = null;

  for (let num of numbers) {
    const remainder = targetSum - num;
    const remainderCombination = bestSum(remainder, numbers);

    if (remainderCombination !== null) {
      const combination = [...remainderCombination, num];
      // if the combination is shorter than the current "shortest" then update it
      if (shortestCombination === null || combination.length < shortestCombination.length) {
        shortestCombination = combination;
      }
    }
  }

  return shortestCombination;
};

// m = targetSum
// n = numbers.length

// Brute force
// time: O(n^m * m)
// space: O(m^2)

// memoised
// time: O(m^2 * n)
// space: O(m^2)

console.log(bestSum(7, [5, 3, 4, 7])); // [7]
console.log(bestSum(8, [2, 3, 5])); // [3, 5]
console.log(bestSum(8, [1, 4, 5])); // [4, 4]
console.log(bestSum(100, [1, 2, 5, 25])); // [25, 25, 25, 25]
//console.log(bestSum(230, [7, 14, -11]));
